go/vt/vtgate/plugin_mysql_server.go

```diff
@@ -436,9 +452,17 @@ func shutdownMysqlProtocolAndDrain() {
 	}
 }
 
+func rollbackAtShutdown() {
```

Comment:
Need to hold mutex here, because you're accessing `connections`. However, you'll run into a deadlock because `ConnectionClosed` also obtains the mutex. 

I recommend implementing a `CloseAllConnections` instead of `ConnectionClosed`, that obtains the mutex and closes all conections in a loop.

And then, you can change `servenv.OnClose(rollbackAtShutdown)` -> `servenv.OnClose(vtgateHandler.CloseAllConnections)`